File: 1901-find-a-peak-element-ii/1901-find-a-peak-element-ii.py
class Solution:
    def findPeakGrid(self, mat: List[List[int]]) -> List[int]:
        def isPeak(x,y,value):
            if x==-1 or y==-1 or x==len(mat) or y==len(mat[0]): return True
            
            return value>mat[x][y]
        
        for i in range(len(mat)):
            for j in range(len(mat[0])):
                value = mat[i][j]
                
                if isPeak(i-1,j,value) and isPeak(i+1,j,value) and isPeak(i,j+1,value) and isPeak(i,j-1,value):
                    return [i,j]
        return [-1,-1]

File: 1901-find-a-peak-element-ii/1901-find-a-peak-element-ii.java
class Solution {
    
    public int getPeakIndex(int[] col) {
        int maxIndex = -1, maxElement = Integer.MIN_VALUE;
        
        for(int i =0; i < col.length; i++) {
            if(maxElement < col[i]) {
                maxIndex = i;
                maxElement = col[i];
            }
        }
        
        return maxIndex;
    }
    
    public int[] findPeakGrid(int[][] mat) {
        int r = mat.length, c = mat[0].length;
        int low = 0, high = r-1;
        
        if(low == high) {
            return new int[]{low, getPeakIndex(mat[0])};
        }
        
        while(low <= high) {
            int mid = (low+high)/2;
            
            int peakIndex = getPeakIndex(mat[mid]);
            
            if(mid < r-1 && mat[mid][peakIndex] < mat[mid+1][peakIndex]) {
                low = mid +1;
            } else if(mid > 0 && mat[mid][peakIndex] < mat[mid-1][peakIndex]) {
                high = mid -1;
            } else{
                return new int[]{mid, peakIndex};
            }
        }
        
        return new int[]{-1,-1};
    }
}